### PR TITLE
Introduce depth scanning feature

### PR DESCRIPTION
**Pull request by Garanas**
This pull request introduces a depth scanning feature. It allows you to know exactly where amphibious units (those that walk over the ocean floor) will go to. Useful when you are trying to interact with entities / resource deposits that are underwater.

https://user-images.githubusercontent.com/15778155/188369339-66042b82-228b-417a-a76c-fb6bb87fc5a2.mp4

### CODE REVIEW

**Comment by Garanas** [comment]
@Hdt80bro could you do a code review when you have the time?

File: lua/ui/game/cursor/depth.lua [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        elevation = cursor.Elevation
+
+        -- check if we have all the data required
+        if position and position[1] and cursor and cursor.Elevation and CheckConditions() then
+            
+            -- move with the grid when building
+            info = CommandMode.GetCommandMode()
+            if info[1] == 'build' then
+                position[1] = (position[1] ^ 0) + 0.5
+                position[3] = (position[3] ^ 0) + 0.5
```
Comment (Hdt80bro) [inline lua/ui/game/cursor/depth.lua:134]: This has the interesting side effect of only showing the depth scanner on the positive portions of out-of-bounds of maps
image

File: lua/ui/game/cursor/depth.lua [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                    bit:SetFractionCompleteParameter(transparency)
+                else
+                    bit:SetHidden(true)
+                end
+            end
+        else
+            -- hide them
+            MeshOnTerrain:SetHidden(true)
+            for k = 1, MeshesInbetweenCount do
+                MeshesInBetween[k]:SetHidden(true)
```
Comment (Hdt80bro) [inline lua/ui/game/cursor/depth.lua:170]: And if performance was *really* an issue, we could localize these upvalues. It's fine the way it is though.

File: lua/ui/game/cursor/depth.lua [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+--- Retrieves cursor information from the engine statistics
+---@return table
+local function GetCursorInformation()
+    local cursor = { }
+    if __EngineStats and __EngineStats.Children then
+        for _, a in __EngineStats.Children do
+            if a.Name == 'Camera' then
+                for _, b in a.Children do
+                    if b.Name == 'Cursor' then
+                        for _, c in b.Children do
```
Comment (Hdt80bro) [inline lua/ui/game/cursor/depth.lua:28]: There's... not a better way to do this is there. At least add some `break`s in.

File: lua/ui/game/cursor/depth.lua [hunk 4]
```diff
@@ -0,0 +1,186 @@
+
+local Prefs = import('/lua/user/prefs.lua')
+local CommandMode = import('/lua/ui/game/commandmode.lua')
+local WorldMesh = import('/lua/ui/controls/worldmesh.lua').WorldMesh
+local Exit = import('/lua/ui/override/Exit.lua')
+
+local meshSphere = '/env/Common/Props/sphere_lod0.scm'
+local meshCircularRing = '/meshes/game/PathRing_LOD0.scm'
+local meshSquareRing = '/meshes/game/PathSquare_LOD0.scm'
+local meshCylinder = '/meshes/game/PathCylinder_LOD0.scm'
```
Comment (Hdt80bro) [inline lua/ui/game/cursor/depth.lua:10]: Can probably remove these unused meshes

**Comment by Hdt80bro** [comment]
It'd probably be more work than it's worth, but `command` mode would be more clear if the scanner only showed up if there was a non-hover amphibious unit in selection.

File: lua/ui/game/cursor/depth.lua [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                MeshOnTerrain:SetStance(location)
+                MeshOnTerrain:SetFractionCompleteParameter(transparency)
+            else
+                MeshOnTerrain:SetHidden(true)
+            end
+
+            -- update visiblity intermediate dots
+            for k = 1, MeshesInbetweenCount do
+                local bit = MeshesInBetween[k]
+                location[2] = (k / 5) * position[2] + (1 - k / 5) * elevation
```
[the remark this answers is not in the document]
Comment (Garanas) [reply lua/ui/game/cursor/depth.lua:155]: That makes it more readable, I'll add it in

File: lua/ui/game/cursor/depth.lua (hunk 2 above)
Comment (Garanas) [reply lua/ui/game/cursor/depth.lua:170]: Yeah, if this was a heavy hitter then I'd agree. I'm not too sure about it in this case, as it has no impact but it does add a few lines of code

File: lua/ui/game/cursor/depth.lua (hunk 3 above)
Comment (Garanas) [reply lua/ui/game/cursor/depth.lua:28]: Not that I am aware of, the statistics have a weird structure that isn't always the same. I'll add in the breaks

**Comment by Garanas** [comment]
> It'd probably be more work than it's worth, but command mode would be more clear if the scanner only showed up if there was a non-hover amphibious unit in selection.

Not intuitive to the player, it may feel 'broken' if they happen to try and scan something quickly. I will be using a similar idea to help people hover bomb (if you have 1 selected unit that is a bomber)

File: lua/ui/game/cursor/depth.lua [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            UniformScale = 0.15,
+            LODCutoff = MeshFadeDistance
+        })
+
+        MeshesInBetween[k] = bit
+        Trash:Add(bit)
+    end
+
+    -- pre-allocate all locals for performance
+    local camera, position, cursor, elevation, transparency, location, info
```
[the remark this answers is not in the document]
Comment (Garanas) [reply lua/ui/game/cursor/depth.lua:118]: It calls it only once though, for all the registers combined.
```
INFO:  - 44: ( 101)   43 - FORLOOP        1  -19
INFO:  - 45: ( 117)   44 - LOADNIL        1    7    0
INFO:  - 46: ( 118)   45 - NEWTABLE       8    0    0
``` 

It is a pattern that I see more often in the base code. I'm not sure if there are benefits to it - I haven't found a way to benchmark it. I think the difference is neglible.

**Pull request merged**